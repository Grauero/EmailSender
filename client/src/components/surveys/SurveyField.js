import React from 'react';

const SurveyField = ({ input, label, meta: { error, touched } }) => (
  <div>
    <label>{label}</label>
    <input {...input} style={{ marginBottom: '5px' }} />
    {touched && error}
  </div>
);

export default SurveyField;
