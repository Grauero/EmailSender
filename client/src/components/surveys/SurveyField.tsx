import React from 'react';

export interface ISurveyField {
  input: object;
  label: string;
  meta: {
    error: string;
    touched: boolean;
  };
}

const SurveyField: React.FC<ISurveyField> = props => {
  const {
    input,
    label,
    meta: { error, touched }
  } = props;

  return (
    <div>
      <label>{label}</label>
      <input {...input} className="input" />
      <div className="red-text touched">{touched && error}</div>
    </div>
  );
};

export default SurveyField;
