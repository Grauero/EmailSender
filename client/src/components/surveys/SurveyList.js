import React, { Component, lazy, Suspense } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Spinner from '../utils/Spinner';
import { fetchSurveys } from '../../store/actions';

const ModalLayout = lazy(() => import('../modal/ModalLayout'));

class SurveyList extends Component {
  state = {
    removingSurvey: null,
    modal: false
  };

  componentDidMount() {
    this.props.fetchSurveys();
  }

  toggleModal = id => {
    this.setState({
      removingSurvey: id,
      modal: !this.state.modal
    });
  };

  renderSurveys = () =>
    this.props.surveys.map(survey => (
      <div className="card survey-card" key={survey._id}>
        <i
          className="material-icons"
          style={{ position: 'absolute', right: 0, cursor: 'pointer' }}
          onClick={() => this.toggleModal(survey._id)}
        >
          close
        </i>
        <div className="card-content">
          <span className="card-title">{survey.title}</span>
          <p>{survey.body}</p>
          <p className="right">
            Sent On: {new Date(survey.dateSent).toLocaleDateString()}
          </p>
        </div>
        <div className="card-action row">
          <span className="green-text text-darken-2 col m12 l6 positive-response">
            Number Of Positive Feedbacks:
            <span className="response"> {survey.yes}</span>
          </span>
          <span className="red-text text-darken-3 col m12 l6 negative-response">
            Number Of Negative Feedbacks:
            <span className="response"> {survey.no}</span>
          </span>
        </div>
      </div>
    ));

  render() {
    const modalLayout = this.state.modal ? (
      <ModalLayout
        id={this.state.removingSurvey}
        toggleModal={this.toggleModal}
      />
    ) : null;

    return (
      <div>
        <Suspense fallback={<Spinner />}>
          {modalLayout}
          {this.renderSurveys()}
        </Suspense>
      </div>
    );
  }
}

SurveyList.propTypes = {
  surveys: PropTypes.instanceOf(Array).isRequired,
  fetchSurveys: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  surveys: state.surveys
});

export default connect(
  mapStateToProps,
  { fetchSurveys }
)(SurveyList);
