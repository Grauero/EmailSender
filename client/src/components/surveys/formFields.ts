export interface IFormField {
  label: string;
  name: string;
}

export const formFields: IFormField[] = [
  { label: 'Survey Title', name: 'title' },
  { label: 'Subject Line', name: 'subject' },
  { label: 'Email Body', name: 'body' },
  { label: 'Recipient List', name: 'recipients' }
];
