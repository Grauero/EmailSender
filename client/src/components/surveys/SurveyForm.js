import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { reduxForm, Field } from 'redux-form';

import SurveyField from './SurveyField';

const SurveyForm = ({ handleSubmit }) => {
  const FIELDS = [
    { label: 'Survey Title', name: 'title' },
    { label: 'Subject Line', name: 'subject' },
    { label: 'Email Body', name: 'body' },
    { label: 'Recipient List', name: 'emails' }
  ];

  function renderFields() {
    return FIELDS.map(({ label, name }) => (
      <Field
        key={name}
        component={SurveyField}
        type="text"
        label={label}
        name={name}
      />
    ));
  }

  return (
    <div>
      <form onSubmit={handleSubmit(values => console.log(values))}>
        {renderFields()}
        <Link to="/surveys" className="red btn-flat white-text">
          Cancel
        </Link>
        <button type="submit" className="teal btn-flat right white-text">
          Next
          <i className="material-icons right">done</i>
        </button>
      </form>
    </div>
  );
};

SurveyForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired
};

export default reduxForm({
  form: 'surveyForm'
})(SurveyForm);
