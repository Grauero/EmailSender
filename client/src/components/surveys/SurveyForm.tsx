import React from 'react';
import { Link } from 'react-router-dom';
import { reduxForm, Field, InjectedFormProps } from 'redux-form';

import SurveyField from './SurveyField';
import validateEmails from '../utils/validateEmails';
import { formFields, IFormField } from './formFields';

export interface IError {
  recipients: string | null;
}

export interface ISurveyForm {
  handleSubmit?: () => void;
  onSurveySubmit: () => void;
}

const SurveyForm: React.FC<InjectedFormProps<{}> & ISurveyForm> = props => {
  const { handleSubmit, onSurveySubmit } = props;

  return (
    <div>
      <form onSubmit={handleSubmit(onSurveySubmit)}>
        {renderFields(formFields)}
        <Link to="/surveys" className="red btn-flat white-text form-back">
          Cancel
        </Link>
        <button
          type="submit"
          className="teal btn-flat right white-text form-submit"
        >
          Next
          <i className="material-icons right">done</i>
        </button>
      </form>
    </div>
  );
};

export function renderFields(formFields: IFormField[]) {
  return formFields.map(({ label, name }) => (
    <Field
      key={name}
      component={SurveyField}
      type="text"
      label={label}
      name={name}
    />
  ));
}

export function validate(values, formFields) {
  const errors: IError = { recipients: null };
  errors.recipients = validateEmails(values.recipients);

  formFields.forEach(({ name }) => {
    if (!values[name]) {
      errors[name] = 'You must provide a value';
    }
  });

  return errors;
}

export { SurveyForm };
export default reduxForm<{}, ISurveyForm>({
  form: 'surveyForm',
  validate,
  destroyOnUnmount: false
})(SurveyForm);
