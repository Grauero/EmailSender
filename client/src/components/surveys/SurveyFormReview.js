import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import formFields from './formFields';

const SurveyFormReview = ({ onCancel, formValues }) => {
  const reviewFields = formFields.map(({ name, label }) => (
    <div key={name}>
      <label>{label}</label>
      <div>{formValues[name]}</div>
    </div>
  ));

  return (
    <div>
      <h5>Confirm your entries</h5>
      {reviewFields}
      <button
        type="button"
        className="yellow darken-3 btn-flat"
        onClick={onCancel}
      >
        Back
      </button>
    </div>
  );
};

SurveyFormReview.propTypes = {
  onCancel: PropTypes.func.isRequired,
  formValues: PropTypes.instanceOf(Object).isRequired
};

const mapStateToProps = state => ({
  formValues: state.form.surveyForm.values
});

export default connect(mapStateToProps)(SurveyFormReview);
