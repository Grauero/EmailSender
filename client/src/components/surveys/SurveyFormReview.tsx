import React from 'react';
import { connect } from 'react-redux';
import { Dispatch, bindActionCreators } from 'redux';
import { withRouter } from 'react-router-dom';

import formFields from './formFields';
import { submitSurvey } from '../../store/actions';
import { IAppState } from '../../store/reducers';

type ReduxProps = ReturnType<typeof mapStateToProps> &
  ReturnType<typeof mapDispatchToProps>;

export interface ISurveyFormReview extends ReduxProps {
  onCancel: () => void;
  history: History;
}

const SurveyFormReview: React.FC<ISurveyFormReview> = props => {
  const { formValues, onCancel, submitSurvey, history } = props;

  const reviewFields = formFields.map(({ name, label }) => (
    <div key={name}>
      <label>{label}</label>
      <div>{formValues ? formValues[name] : ''}</div>
    </div>
  ));

  return (
    <div className="preview">
      <h5>Confirm your entries</h5>
      {reviewFields}
      <button
        type="button"
        className="yellow darken-3 white-text btn-flat form-back"
        onClick={onCancel}
      >
        Back
      </button>
      <button
        type="button"
        className="green btn-flat right white-text form-submit"
        onClick={() => submitSurvey(formValues, history)}
      >
        Send Survey
        <i className="material-icons right">email</i>
      </button>
    </div>
  );
};

const mapStateToProps = (state: IAppState) => ({
  formValues: state.form.surveyForm.values
});

const mapDispatchToProps = (dispatch: Dispatch) => {
  return bindActionCreators({ submitSurvey }, dispatch);
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(SurveyFormReview));
