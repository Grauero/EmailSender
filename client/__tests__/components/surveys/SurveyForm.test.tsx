import 'jest';
import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import {
  renderFields,
  validate,
  SurveyForm,
  ISurveyForm
} from '../../../src/components/surveys/SurveyForm';
import { formFields } from '../../../src/components/surveys/formFields';
import { InjectedFormProps } from 'redux-form';

const props = ({
  onSurveySubmit: jest.fn(),
  handleSubmit: jest.fn()
} as unknown) as ISurveyForm & InjectedFormProps;
const component = shallow(<SurveyForm {...props} />);

test('renderFields function returns correct number of <Field> components', () => {
  const zeroItems = [];
  const oneItem = [{ label: 'l1', name: 'n1' }];
  const twoItems = [{ label: 'l1', name: 'n1' }, { label: 'l2', name: 'n2' }];

  expect(renderFields(zeroItems)).toMatchSnapshot();
  expect(renderFields(oneItem)).toMatchSnapshot();
  expect(renderFields(twoItems)).toMatchSnapshot();
});

test('validate function validates provided data and returns NULL if data is valid', () => {
  const values = {
    recipients: '1@gmail.com',
    title: 'title',
    body: 'body',
    subject: 'subject'
  };

  expect(validate(values, formFields)).toEqual({ recipients: null });
});

test('validate function validates provided data and return error object if data is invalid', () => {
  const values = {};

  expect(validate(values, formFields)).toEqual({
    body: 'You must provide a value',
    recipients: 'You must provide a value',
    subject: 'You must provide a value',
    title: 'You must provide a value'
  });
});

test('matches snapshot', () => {
  expect(toJSON(component)).toMatchSnapshot();
});
