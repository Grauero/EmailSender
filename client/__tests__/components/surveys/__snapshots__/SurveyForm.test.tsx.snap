// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <form>
    <Field
      component={[Function]}
      key="title"
      label="Survey Title"
      name="title"
      type="text"
    />
    <Field
      component={[Function]}
      key="subject"
      label="Subject Line"
      name="subject"
      type="text"
    />
    <Field
      component={[Function]}
      key="body"
      label="Email Body"
      name="body"
      type="text"
    />
    <Field
      component={[Function]}
      key="recipients"
      label="Recipient List"
      name="recipients"
      type="text"
    />
    <Link
      className="red btn-flat white-text form-back"
      replace={false}
      to="/surveys"
    >
      Cancel
    </Link>
    <button
      className="teal btn-flat right white-text form-submit"
      type="submit"
    >
      Next
      <i
        className="material-icons right"
      >
        done
      </i>
    </button>
  </form>
</div>
`;

exports[`renderFields function returns correct number of <Field> components 1`] = `Array []`;

exports[`renderFields function returns correct number of <Field> components 2`] = `
Array [
  <Field
    component={[Function]}
    label="l1"
    name="n1"
    type="text"
  />,
]
`;

exports[`renderFields function returns correct number of <Field> components 3`] = `
Array [
  <Field
    component={[Function]}
    label="l1"
    name="n1"
    type="text"
  />,
  <Field
    component={[Function]}
    label="l2"
    name="n2"
    type="text"
  />,
]
`;
